File: sdks/js/src/compat.ts
/** Unstructured API compatibility — drop-in replacement for unstructured-client. */

import type { Element } from "./types.js";
import { DocParseError, AuthError, QuotaError } from "./types.js";

/** Detect auth-related error messages from server-side envelope errors. */
function isAuthErrorMessage(msg: string): boolean {
  if (!msg) return false;
  const m = msg.toLowerCase();
  return (
    m.includes("invalid or expired api key") ||
    m.includes("invalid api key") ||
    m.includes("missing api key") ||
    m.includes("unauthorized") ||
    m.includes("api key required")
  );
}

function raiseEnvelopeError(msg: string): never {
  if (isAuthErrorMessage(msg)) throw new AuthError(msg);
  throw new DocParseError(msg);
}

const DEFAULT_BASE_URL = "https://docparse.ailang.sunholo.com";

class GeneralApi {
  private baseUrl: string;
  private apiKey: string;
  private timeout: number;

  constructor(baseUrl: string, apiKey: string, timeout: number) {
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
    this.timeout = timeout;
  }

  /**
   * Partition a document — returns Unstructured-format elements.
   *
   * Accepts a local file path (uploaded via multipart, Node.js only) or a
   * sample ID (sent as JSON). Usage is identical to unstructured-client.
   */
  async partition(opts: { file: string; strategy?: string }): Promise<Element[]> {
    const url = `${this.baseUrl}/general/v0/general`;
    const headers: Record<string, string> = {};
    if (this.apiKey) {
      headers["unstructured-api-key"] = this.apiKey;
    }

    let resp: Response;
    // Detect Node.js vs browser for local file upload
    const isNode = typeof process !== "undefined" && process.versions?.node;
    let isLocalFile = false;
    if (isNode) {
      const { existsSync, statSync } = await import("fs");
      isLocalFile = existsSync(opts.file) && statSync(opts.file).isFile();
    }

    if (isLocalFile) {
      const { readFileSync } = await import("fs");
      const { basename } = await import("path");
      const fileData = readFileSync(opts.file);
      const blob = new Blob([fileData]);
      const form = new FormData();
      form.append("files", blob, basename(opts.file));
      if (opts.strategy) form.append("strategy", opts.strategy);
      resp = await fetch(url, { method: "POST", headers, body: form });
    } else {
      headers["Content-Type"] = "application/json";
      resp = await fetch(url, {
        method: "POST",
        headers,
        body: JSON.stringify({ filepath: opts.file, strategy: opts.strategy || "auto" }),
      });
    }

    if (resp.status === 401) throw new AuthError();
    if (resp.status === 429) throw new QuotaError("Quota exceeded");
    if (!resp.ok) throw new DocParseError(`API error: ${resp.status}`, resp.status);

    const outer = await resp.json();
    if (outer.error && typeof outer.error === "string") {
      raiseEnvelopeError(outer.error);
    }

    const resultStr = outer.result || "[]";
    try {
      const parsed = JSON.parse(resultStr);
      // Check for error in inner result
      if (parsed?.error) {
        const err = parsed.error;
        const msg = typeof err === "object" ? err.message || JSON.stringify(err) : String(err);
        raiseEnvelopeError(msg);
      }
      return Array.isArray(parsed) ? parsed : [];
    } catch (e) {
      if (e instanceof DocParseError) throw e;
      return [];
    }
  }
}

/**
 * Drop-in replacement for `unstructured-client`'s UnstructuredClient.
 *
 * Migration:
 * ```typescript
 * // Before
 * import { UnstructuredClient } from 'unstructured-client';
 * const client = new UnstructuredClient({ serverUrl: 'https://api.unstructured.io' });
 *
 * // After — one import change
 * import { UnstructuredClient } from '@ailang/parse/compat';
 * const client = new UnstructuredClient({
 *   serverUrl: 'https://api.parse.sunholo.com'
 * });
 * ```
 */
export class UnstructuredClient {
  general: GeneralApi;

  constructor(opts: { serverUrl?: string; apiKey?: string; timeout?: number }) {
    const baseUrl = (opts.serverUrl || DEFAULT_BASE_URL).replace(/\/$/, "");
    // Resolve key: explicit > env var
    let key = opts.apiKey || "";
    if (!key) {
      try { key = process.env.DOCPARSE_API_KEY || ""; } catch { /* browser */ }
    }
    this.general = new GeneralApi(baseUrl, key, opts.timeout || 60000);
  }
}

File: sdks/r/R/compat.R
#' Unstructured.io API compatibility layer
#'
#' Drop-in replacement for the \code{UnstructuredClient} class shipped by
#' \pkg{unstructured-client} (Python) and similar wrappers. Posts to
#' \code{/general/v0/general} on the configured server, returning a list of
#' \code{Element} records with the same field names as the Unstructured API.
#'
#' Migration from a hypothetical \code{unstructured} R wrapper:
#' \preformatted{
#' uc <- ailangparse::UnstructuredClient$new(
#'   server_url = "https://docparse.ailang.sunholo.com"
#' )
#' elements <- uc$partition("report.docx", strategy = "auto")
#' }
#'
#' @examples
#' \dontrun{
#' uc <- UnstructuredClient$new()
#' uc$partition("report.docx", strategy = "auto")
#' }
#' @export
UnstructuredClient <- R6::R6Class(
  "UnstructuredClient",
  public = list(
    #' @field server_url Base URL of the Unstructured-compatible endpoint.
    server_url = NULL,
    #' @field api_key Active API key, or \code{""}.
    api_key = NULL,
    #' @field timeout Per-request timeout in seconds.
    timeout = NULL,

    #' @description Construct a new client.
    #' @param server_url Base URL of the Unstructured-compatible endpoint.
    #' @param api_key Optional API key. If empty, the env var
    #'   \code{DOCPARSE_API_KEY} and saved credentials are consulted.
    #' @param timeout Per-request timeout in seconds.
    initialize = function(server_url = default_base_url(),
                          api_key = NULL,
                          timeout = 60) {
      self$server_url <- sub("/$", "", server_url)
      self$timeout    <- timeout
      key <- if (is.null(api_key)) "" else api_key
      if (!nzchar(key)) {
        key <- Sys.getenv("DOCPARSE_API_KEY", unset = "")
      }
      if (!nzchar(key)) {
        saved <- load_saved_key(self$server_url)
        if (!is.null(saved)) key <- saved$api_key
      }
      self$api_key <- key
      invisible(self)
    },

    #' @description Partition a document into Unstructured-style elements.
    #' @param file Local file path (uploaded multipart) or sample ID /
    #'   server-side filepath (sent as JSON).
    #' @param strategy Partitioning strategy passed to the API
    #'   (default \code{"auto"}).
    #' @return A list of element records (each a list with \code{type},
    #'   \code{element_id}, \code{text}, \code{metadata}).
    partition = function(file, strategy = "auto") {
      url_path <- "/general/v0/general"
      req <- httr2::request(paste0(self$server_url, url_path))
      req <- httr2::req_timeout(req, self$timeout)
      req <- httr2::req_user_agent(req, "ailangparse-r/0.4.1")
      req <- httr2::req_error(req, is_error = function(resp) FALSE)
      if (nzchar(self$api_key)) {
        req <- httr2::req_headers(req, `unstructured-api-key` = self$api_key)
      }

      if (file.exists(file)) {
        req <- httr2::req_body_multipart(
          req,
          files    = curl::form_file(file, name = basename(file)),
          strategy = strategy
        )
      } else {
        req <- httr2::req_body_json(
          req,
          list(
            filepath = jsonlite::unbox(file),
            strategy = jsonlite::unbox(strategy)
          )
        )
      }

      body_text <- .perform(req)
      outer <- tryCatch(
        jsonlite::fromJSON(body_text, simplifyVector = FALSE),
        error = function(e) {
          stop(.docparse_error(sprintf("Invalid JSON response: %s",
                                       conditionMessage(e))))
        }
      )
      if (!is.null(outer$error) && length(outer$error) > 0L) {
        stop(.docparse_error(as.character(outer$error)))
      }
      result_str <- if (is.null(outer$result)) "[]" else outer$result
      elements_raw <- tryCatch(
        jsonlite::fromJSON(result_str, simplifyVector = FALSE),
        error = function(e) list()
      )
      if (is.list(elements_raw) && !is.null(elements_raw$error)) {
        err <- elements_raw$error
        msg <- if (is.list(err) && !is.null(err$message)) err$message
               else as.character(err)
        stop(.docparse_error(msg))
      }
      if (!is.list(elements_raw)) return(list())
      lapply(elements_raw, .element_from_list)
    }
  )
)
